secure_software_development_fundamentals.md

```diff
@@ -4812,7 +4814,13 @@ Whenever you compare secret values or cryptographic values (such as session keys
 
 #### Minimizing the Time Keys/Decrypted Data Exists
 
-Remember that per least privilege, we want to minimize the time a privilege is active. In cryptography, you often want to minimize the time a private key or password is available, or at least minimize the time that the decrypted data is available. This can be harder that you might think. At the operating system level you can probably lock it into memory with **mlock()** or **VirtualLock()**; this will at least prevent the data from being copied into storage. Ideally, you would erase it from memory after use, though that is often surprisingly difficult. Compilers may turn overwrite code into a no-op, because they detect that nothing reads the overwritten values. Languages with built-in garbage collection often quietly make extra copies and/or do not provide a mechanism for erasure. That said, some languages or infrastructure do make this easy. For example, those using the .NET framework (e.g., C#) can use SecureString.
+Remember that per least privilege, we want to minimize the time a privilege is active. In cryptography, you often want to minimize the time a private key or password is available, or at least minimize the time that the decrypted data is available. This can be harder that you might think. At the operating system level you can probably lock it into memory with **mlock()** or **VirtualLock()**; this will at least prevent the data from being copied into storage. Ideally, you would erase it from memory after use using interfaces that are safe for that purpose. The reason is that compilers since they detect that nothing reads the overwritten values optimize the code out as no-op. The following interfaces in C and C++ are safe for overwriting sensitive data in memory.
```

Comment:
> The reason is that compilers since they detect that nothing reads the overwritten values optimize the code out as no-op.

This sentence doesn't make sense. Did you mean:

> Compiler optimizers may quietly eliminate the code to overwrite data, because they detect that nothing else in the program reads the overwritten values. In addition, languages with built-in garbage collection often quietly make extra copies and/or do not provide a mechanism for erasure.

Also, change:

> safe for overwriting

to:

> safer for overwriting

as there are still ways to misuse them, esp. if data regions are declared as volatile. See:
* https://www.gnu.org/software/gnulib/manual/html_node/explicit_005fbzero.html
* https://www.gnu.org/software/gnulib/manual/html_node/memset_005fexplicit.html

It doesn't mention `memset_explicit`, but that's new so I think that's okay.

Reply:
I couldn't have expressed it better. Thanks, I've updated the text.

I checked GNU [libc manual](https://sourceware.org/glibc/manual/latest/html_mono/libc.html) and there is only `explicit_bzero`. `explicit_memset` seems to be available only in gnulib (which despite the name is not the same as the gnu libc).

Reply:
The text ` In addition, languages with built-in garbage collection often quietly make extra copies and/or do not provide a mechanism for erasure.` was listed at the end, after the C/C++ interfaces. I merged the texts, and kept `For example, those using the .NET framework (e.g., C#) can use SecureString.`. 

As a result the C/C++ interfaces are now listed at the end.